# Исследование надежности заемщиков


**Описание данных:**
1. children — количество детей в семье
2. days_employed — общий трудовой стаж в днях
3. dob_years — возраст клиента в годах
4. education — уровень образования клиента
5. education_id — идентификатор уровня образования
6. family_status — семейное положение
7. family_status_id — идентификатор семейного положения
8. gender — пол клиента
9. income_type — тип занятости
10. debt — имел ли задолженность по возврату кредитов
11. total_income — ежемесячный доход
12. purpose — цель получения кредита


## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# создаем функцию, которая словестно будет выводить количество детей у каждого кредитора
#(делается для более приятного просмотра и понятного результата)
def number_of_children(numder):
    try:
        if numder == 0:
            return 'Детей нет'
        elif numder == 1:
            return 'Один ребенок'
        elif numder == 2:
            return 'Два ребенка'
        elif numder >= 3:
            return 'Три и более детей'
    except:
        pass
    
#создаем новый столбец, используя ранее созданную функцию
data['number_of_children'] = data['children'].apply(number_of_children)
#создаем сводную таблицу с общим числом кредиторов и числом кредиторов с задолжностью, 
#подсчитанные по каждой группе 'number_of_children'
data_pivot_1 = data.pivot_table(index = ['number_of_children'], values = ['debt'], aggfunc = ['count', 'sum'])
#создаем новый столбец для подсчета процентного соотношения кредиторов
#с задолжностью к общему числу кредиторов
data_pivot_1['ratio'] = (data_pivot_1['sum'] / data_pivot_1['count'])
data_pivot_1['ratio'] = data_pivot_1['ratio'].transform(lambda x: '{:,.2%}'.format(x))
#отсортируем таблицу по столбцу 'ratio'
data_pivot_1.sort_values(by = 'ratio')

count   sum  ratio
                     debt  debt       
number_of_children                    
Детей нет           14091  1063  7.54%
Три и более детей     380    31  8.16%
Один ребенок         4808   444  9.23%
Два ребенка          2052   194  9.45%

**Вывод:** Кредиторы, не имеющие детей наиболее чаще погашают кредиты в срок, как и семьи с тремя и более детьми. В свою очередь кредиторы, имеющие одного или двоих детей чаще имеют просрочку по кредиту. Поэтому сделать вывод, что чем больше детей, тем выше вероятность погашения кредита не в срок, на основе сводной таблицы невозможно.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
#создаем сводную таблицу с общим числом кредиторов и числом кредиторов с задолжностью, 
#подсчитанные по каждой группе 'family_status'
data.pivot_table(index = ['family_status'], values = ['debt'], aggfunc = ['count', 'sum'])
data_pivot_2 = data.pivot_table(index = ['family_status'], values = ['debt'], aggfunc = ['count', 'sum'])
#создаем новый столбец для подсчета процентного соотношения кредиторов
#с задолжностью к общему числу кредиторов
data_pivot_2['ratio'] = (data_pivot_2['sum'] / data_pivot_2['count'])
data_pivot_2['ratio'] = data_pivot_2['ratio'].transform(lambda x: '{:,.2%}'.format(x))
#отсортируем таблицу по столбцу 'ratio'
data_pivot_2.sort_values(by = 'ratio')

count  sum  ratio
                        debt debt       
family_status                           
вдовец / вдова           951   63  6.62%
в разводе               1189   84  7.06%
женат / замужем        12261  927  7.56%
гражданский брак        4134  385  9.31%
Не женат / не замужем   2796  273  9.76%

**Вывод:** Наиболее низкие риски невозврата в срок кредита имеют вдовцы/вдовы, а также люди, находящиеся в разводе. А не женатые/не замужние кредиторы представляют наибольшую вероятность возврата кредита не в срок.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
#создаем сводную таблицу с общим числом кредиторов и числом кредиторов с задолжностью, 
#подсчитанные по каждой группе 'total_income_category'
data.pivot_table(index = ['total_income_category'], values = ['debt'], aggfunc = ['count', 'sum'])
data_pivot_3 = data.pivot_table(index = ['total_income_category'], values = ['debt'], aggfunc = ['count', 'sum'])
#создаем новый столбец для подсчета процентного соотношения кредиторов
#с задолжностью к общему числу кредиторов
data_pivot_3['ratio'] = (data_pivot_3['sum'] / data_pivot_3['count'])
data_pivot_3['ratio'] = data_pivot_3['ratio'].transform(lambda x: '{:,.2%}'.format(x))
#отсортируем таблицу по столбцу 'ratio'
data_pivot_3.sort_values(by = 'ratio')



count   sum  ratio
                        debt  debt       
total_income_category                    
D                        349    21  6.02%
B                       5014   354  7.06%
A                         25     2  8.00%
C                      15921  1353  8.50%
E                         22     2  9.09%

**Вывод:** Наиболее низкие риски невозврата кредита имеют кредиторы, по уровню дохода относящиеся к группе D(30001–50000). Кредиторы группы E(0–30000) представляют наибольшую вероятность возврата кредита не в срок. Утверждать, что с увелечением дохода риски невозврата кредита в срок уменьшаются, нельзя.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
#создаем сводную таблицу с общим числом кредиторов и числом кредиторов с задолжностью, 
#подсчитанные по каждой группе 'purpose_category'
data.pivot_table(index = ['purpose_category'], values = ['debt'], aggfunc = ['count', 'sum'])
data_pivot_4 = data.pivot_table(index = ['purpose_category'], values = ['debt'], aggfunc = ['count', 'sum'])
#создаем новый столбец для подсчета процентного соотношения кредиторов
#с задолжностью к общему числу кредиторов
data_pivot_4['ratio'] = (data_pivot_4['sum'] / data_pivot_4['count'])
data_pivot_4['ratio'] = data_pivot_4['ratio'].transform(lambda x: '{:,.2%}'.format(x))
#отсортируем таблицу по столбцу 'ratio'
data_pivot_4.sort_values(by = 'ratio')

count  sum  ratio
                           debt debt       
purpose_category                           
операции с недвижимостью  10751  780  7.26%
проведение свадьбы         2313  183  7.91%
получение образования      3988  369  9.25%
операции с автомобилем     4279  400  9.35%

**Вывод:** Люди, берущие кредит для операций с недвижимостью чаще возвращают его в срок, чем те, кто берет кредит на получение образования или на операциии с автомобилем. Следовательно, выгоднее всего выдавать кредиты на цели, связанные с недвижимостью.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Пропуски могли появиться при выгрузке данных, ошибках в логике подсчета данных или данные по тем или иным причинам были пропущены или отсутствовали изначально.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства (в выборке присутствуют выдающиеся значения: 500 против 200), для таких случаев лучше использовать метод median() для заполнения пропусков.

### Шаг 4: общий вывод.

В ходе исследования надёжности заёмщиков были проведены следующие расчеты:

**1) Зависимость между количеством детей и вероятности возврата кредита не в срок:**

- Детей нет - 7.54%
- Три и более детей - 8.16%
- Один ребенок - 9.23%
- Два ребенка - 9.45%

Абсолютная разница составила: 1.91%

Следовательно, наиболее надежный заемщик это человек, не имеющий детей, так как именно у этой группы заемщиков вероятность возврата выданного банком кредита вовремя наиболее высока. В ходе анализа прямой зависимости влияния количества детей в семье на возврат кредита в срок не было найдено.

**2) Зависимость между семейным положением и вероятности возврата кредита не в срок:**

- Вдовец/Вдова - 6.62%
- В разводе - 7.06%
- Женат/Замужем - 7.56%
- Гражданский брак - 9.31%
- Не женат/Не замужем - 9.76%

Абсолютная разница составила: 3.14%

Следовательно, наиболее надежный заемщик это вдовец/вдова, именно у этой группы заемщиков вероятность возврата выданного банком кредита вовремя наиболее высока.

**3) Зависимость между уровнем дохода и вероятности возврата кредита не в срок:**

- Группа D(доход: 30.001-50.000) - 6.02%
- Группа B(доход: 200.000-1.000.000) - 7.06%
- Группа A(доход: 1.000.001 и выше) - 8.00%
- Группа C(доход: 50.001-200.000) - 8.50%
- Группа E(доход: 0-30.000) - 9.09%

Абсолютная разница составила: 3.07%

Следовательно, наиболее надежный заемщик это человек, чей доход находится в диапазоне от 30.001 до 50.000, у этой группы заемщиков вероятность возврата выданного банком кредита вовремя наиболее высока. В ходе анализа прямой зависимости влияния уровня дохода кредитора на возврат кредита в срок не было найдено.

**4) Влияние целей кредита на вероятность возврата кредита не в срок:**

- Операции с недвижимостью - 7.25%
- Проведения свадьбы - 7.91%
- Получение образования - 9.25%
- Операции с автомобилем - 9.35%

Абсолютная разница составила: 2.1%

Следовательно, наиболее надежный заемщик это человек, которому кредит необходим для совершения операций с недвижимостью, вероятность возврата выданного банком кредита вовремя у этой группы наиболее высока.

**В ходе анализа можно сделать вывод, что наиболее важным критерием, на который стоит обращать внимание банкам перед выдачей кредита - это семейное положение кредитора, далее за ним идут его доход и цель получения кредита, а последнее, на что стоит обращать внимание, - это наличие детей.**

Также можно составить "портрет" наиболее надежного кредитора банка: вдовец/вдова без детей, с уровнем дохода в 30001 - 50000 (относящийся к группе D), которому кредит необходим для совершения операциий с недвижимостью. И наименее надежным кредитором банка будет не женатый/не замужний человек с двумя детьми, с уровнем дохода 0-30000 (относящийся к группе E), который берет кредит для операций с автомобилем.
